src/GLSL.hs

```diff
@@ -0,0 +1,75 @@
+module GLSL where
+import Syntax (OpaqueType)
+
+data GSize = Two
+           | Three
+           | Four
+        deriving (Eq, Ord)
+
+data GType = Void
+           | GBool
+           | GInt
+           | GUInt
+           | GFloat
+           | GDouble
+           | GVec GSize GType
+           | GMat GSize GSize
+
+data GOp = GArr
+         | GField
+         | GMult
+         | GMod
+         | GDiv
+         | GAdd
+         | GSub
+         | GRShift
+         | GLShift
+         | GLt
+         | GLte
+         | GGt
+         | GGte
+         | GEq
+         | GNEq
+         | GAnd
+         | GOr
+         | GXOr
+         | GBAnd
+         | GBOr
+         | GBXOr
+
+data GUOp = GFLip
+          | GNot
+          | GNeg
+          | GPos
+
+data GConst = GR String
+            | GI Int
+            | GB Bool
+            | GF Float
+            | GD Double
+            | GV [GConst]
+            | GM [[GConst]]
+
+data GExpr = GCall String [GExpr]
+           | GParen GExpr
+           | GAssign String GExpr
+           | Imm GConst
+           | GUn GUOp GExpr
+           | GBin GOp GExpr GExpr
+           | GTern GExpr GExpr GExpr
+
+data GStmt = GFor GExpr GExpr GExpr [GStmt]
+           | GWhile GExpr [GStmt]
+           | GDoWhile GExpr [GStmt]
+           | GIf [(Maybe GExpr, [GStmt])] -- This is so that we can have if/else if/else chains
```

Comment:
We should enforce that if & else-if must always have a condition, and must always have at least one branch (possibly followed by a branch w/ no condition)
```haskell
type CondBranch = (Cond,[Stmt])

-- a bit crude, but then we have the idea that a
-- branch is minimum 1 branch w/ statements guarded by a condition,
-- followed by zero or more condtional branches & maybe a final `else`
data GStmt = ...
     | GIf CondBranch [CondBranch] (Maybe [Stmt])
```

Reply:
Aaaah very clever and  a more correct way to handle this. Though here you don't define `Cond`. Should this be a lower nonterminal than expression?